### ISSUE TITLE
Is there a better way to track time dependent information in DQM?

### ISSUE THREAD

**Issue by davidlange6** [cms-sw/cmssw#19270]
We are getting requests for a lot of new histograms with ~2500 bins on the x axis (sometimes profiles, sometimes 2d) whose purpose is to monitor things vs lumi section.  (2500 corresponds to a 16 hour run...) Eg, I think one PR has 500 (y bins) x2500 (lumis)x 10 (instances) [then x8 cores = 100MB)

While perhaps one job will process an entire run in the online, I would think no more than a handful (or two) lumi sections are processed in a prompt reco job. (so >25x too much memory being allocated)

So, do we not have a way to move this memory expensive construct to where the statistics for a run are aggregated together? If not are there thoughts on how to make one?

**Comment by davidlange6** [comment]
assign dqm

**Comment by davidlange6** [comment]
assign core

**Comment by Dr15Jones** [comment]
@davidlange6 I would think that the aggregation across lumis could be done in the Harvesting step.

**Comment by davidlange6** [comment]
can you think of an example to point people towards?

**Comment by Dr15Jones** [comment]
I have no idea if anyone has done such a thing, I was only saying that I could image one could do it. The Harvesting step allows someone to read MonitorElements from the DQMStore and then create new MonitorElements based on those previous ones. Off the top of my head I can't think of why the read MonitorElements couldn't come from the LuminosityBlock transition and then the final MonitorElement is finished at end job time (which is where the DQMEDHarvester does its final call for MonitorElements).

**Comment by VinInn** [comment]
I think we should ask root to implement histograms that do not allocate (interval of) bins if they are not filled. can be done under the hood...
(I book 0,2500, I use only 11 it books only one bin corresponding to 11, etc)
when are summed the whole interval (say 9 250) is properly filled....
pretty trivial

**Comment by davidlange6** [comment]
a good thought -I realize  its already done for the example of lumi section # if we know the lumi section number at booking time - make a one bin histogram and set SetCanExtend(). but can the lumi section be known when the histrograms are booked? I would think not..

**Comment by davidlange6** [comment]
or has anyone used THnSparse?

**Comment by davidlange6** [comment]
Ignore SetCanExtend - it doesn't change the # of bins when extending the axis...

**Comment by VinInn** [comment]
otherwise do-it-yourself
a map (with one single entry most of the time) of single bin histos
and fill the proper histo in harvesting
A waste of intelligence, ok, but is ROOT developers' intelingence...

**Comment by venturia** [comment]
@davidlange6 "ignore SetCanExtend" if you want to be sure about having one bin for each lumisection BUT, on the other hand, if you want to have trend plots of average values where one bin is one or more (but not too many) lumisections, SetCanExtend is the right thing to do. This is what we are doing in the Tracker and we are pretty happy about them. SetCanExtend does everything for us: change the binning when the number of lumisections pass a threshold value and handles properly the mering of the histograms (TProfile) even if they end up having different number of bins.

**Comment by davidlange6** [comment]
interesting - perhaps it works differently for TProfiles than TH1Fs (in fact, probably it has to be the case given what a TProfile does) - will try to confirm. 

It still doesn't suppress the 0 on the x-axis automatically however... so if you are processing lumi section 25000, you have 25000 bins..

**Comment by venturia** [comment]
OK, but we do not book with 25000 bins... One has to be much much less demanding.

**Comment by davidlange6** [comment]
indeed, i was exaggerating by a factor of 10 - but 2500 is routine.

**Comment by venturia** [comment]
2500 seems to be one bin per LS for runs long as much as ~16 hours. It looks conservative to me. I see no problem to go to a factor 4 less for plots that do not require, strictly, "one LS in one bin".

**Comment by fwyzard** [comment]
The "waste" (whether we have 2500 or a factor 4 less bins) is that the Prompt and Express jobs usual work on few lumisections at a time.
So a sparse histogram storing (and merging) only the non-empty bins would save at least a factor 100.

**Comment by dmitrijus** [comment]
'SetCanExtend' is extremely problematic for DQM.
It makes merging histograms very slow, TH1::Add simply rejects such histograms. 

This is the 'merging' which is done at several levels: between two threads, then opening files for harvesting and for merging histograms coming out of Online HLT. If you set 'kCanExtend', you somewhat destroy the performance of all of those.

The same exact worry applies to THnSparse. It supports THnSparse::Add, but the implementation seems somewhat slow. Can not say by how much.

Also, THnSparse uses more memory and is much slower at pretty much every operation. If you fill 1/100 of it, it might be faster, but once it is fully filled - it will be much slower and take at least twice as much memory as the regular TH1. 

Even bigger issue is that it makes memory usage no longer fixed. TH1 allocates everything once the booking is done, so if cmssw decides to crash due to maxRSS, it will do so within first event.
If we go for THnSparse, the memory grows over time and the crash no longer happens at the convenient first event.

**Comment by davidlange6** [comment]
> On Jun 19, 2017, at 5:42 PM, Dmitrijus <notifications@github.com> wrote:
> 
> 'SetCanExtend' is extremely problematic for DQM.
> It makes merging histograms very slow, TH1::Add simply rejects such histograms.
> 
> This is the 'merging' which is done at several levels: between two threads, then opening files for harvesting and for merging histograms coming out of Online HLT. If you set 'kCanExtend', you somewhat destroy the performance of all of those.
> 
> The same exact worry applies to THnSparse. It supports THnSparse::Add, but the implementation seems somewhat slow. Can not say by how much.
> 
> Also, THnSparse uses more memory and is much slower at pretty much every operation. If you fill 1/100 of it, it might be faster, but once it is fully filled - it will be much slower and take at least twice as much memory as the regular TH1.

right, so don't fully fill - merge into a TH1F/TH2F..  Shall be interesting to benchmark the two. Can DQM support a THnSparse?



> 

> Even bigger issue is that it makes memory usage no longer fixed. TH1 allocates everything once the booking is done, so if cmssw decides to crash due to maxRSS, it will do so within first event.
> If we go for THnSparse, the memory grows over time and the crash no longer happens at the convenient first event.
> 

crashes in the real world CMS don't happen on the "first event". They happen far into the job as memory creeps up (or on busy events).

**Comment by fwyzard** [comment]
An alternative would be to roll our own sparse histogram class (and possibly contribute it back to root).
For example, reading the documentation and that of THnSparse, Eigen sparse matrices look to be much more optimised memory-wise, and could be converted to non-sparse matrices once they are filled above a given threshold.

But we do need somebody to work on it, and benchmark it all..

P.S.
loosely related, I've started to play with "histograms" in multiple threads using atomic vs mutexes

**Comment by Dr15Jones** [comment]
Jim Pivarski did a study and showed atomics are a good choice. He had a talk in Core.

**Comment by fwyzard** [comment]
Do you have a link ?

**Comment by Dr15Jones** [comment]
I was wrong, Jim only send that work to me. @jpivarski would you be willing to share your concurrent histogram filling results here?

**Comment by jpivarski** [comment]
I did a study of dense, concurrent histograms https://indico.cern.ch/event/607830/contributions/2613395/ to entice someone to implement them as a part of ROOT.

What I found is that copy-and-swap and std::atomic algorithms scale extremely well— the horizontal axis is dozens of threads— and that the failure to scale is dominated by something other than contention (two threads wanting to increment the same bin at the same time). We know this because the scaling of a naive implementation (hist[i]++) is not much better. At these GHz rates, it's probably a hardware limitation— serializing on the memory bus or copying from CPU cache to main memory.

Sparse histograms would be a different story, because that's more like a hash map than an array. But given the success of std::atomic[NUM_BINS], it would be worth trying std::map<long, std::atomic<long>>. Or the copy-and-swap method, because the you can see what's going on and measure collision rate. (std::atomic mysteriously does _better_ than the naive implementation in some cases, and not understanding that makes me wary).

Oh, and I've put all the code for these studies on https://github.com/diana-hep/parallel-histogram-metrics . Feel free to fork it.

**Comment by davidlange6** [comment]
I had a look at using THnSparse vs TH2F. 

I made a giant 2D histogram 2500 in x (eg, lumi sections) and 10k in y (random variable to plot). Then I filled the plots one lumi section at a time (100k times per "lumi section"). I did this for a number of lumi sections (either 4 or 1200). Finally I added histograms one by one to get the total sum (to simulate merging results across jobs)

When using 4 "lumi sections"
 - THnSparse is considerably faster. (the accuracy of my measurement is not great, but something like 6x)
 - THnSparse has much less memory allocations (new delete a histogram for each lumi)

When using 1200 "lumi sections"
 - THnSparse is considerably faster. (40 ticks vs 220 ticks)
 - THnSparse has much less memory allocations (new delete a histogram for each lumi)
 - THnSparse takes much more memory for the final histogram (9x!)

So, bugs aside, assuming not that many lumis are processed in one job and a not-too-memory hungry merge, this works pretty well. (faster than what we currently do). I agree that other solutions could be better - including eigen, or some data structure that knew we were sparse in one axis and not the others...

**Comment by dmitrijus** [comment]
> Can DQM support a THnSparse?
We can and it would most definitely solve Online HLT case, where only one bin per lumi is filled (in a lumi-binned histogram). My biggest worry is simply misuse. Immediately, after we enable it, people will start using it for everything and in ways we have not yet imagined.

In fact, this entire case we are discussing is a misinterpretation of what a histogram is. Since we are no longer interested in the distribution of some variable, but a variable's evolution over time (= time series).

I would love to take a look at THnSparse, but unfortunately I am too busy this week. I can't even review the pull requests properly...

Atomic are great to see, but they have to be supported by and integrated into root's TH1. All design decisions of DQM were around TH1 and it's features. Rolling our own implementation might do the trick, but there is simply too much code which relies on TH1 logic. A lot of histograms are not 'standard' histograms too, they are populated with setBinContent() rather than Fill()

**Comment by fwyzard** [comment]
I opened #19395 for the discussion about concurrency.

**Comment by dmitrijus** [comment]
@davidlange6 Thanks for the test.

There is an ongoing project to add TH2Poly into DQM supported types, so I am adding THnSparseD/F to that list.

**Comment by smuzaffar** [comment]
@davidlange6 , can we close this issue?

**Comment by davidlange6** [comment]
I’m not aware of a fix/conclusion here

**Comment by smuzaffar** [comment]
I guess this is still not fixed/resolved

**Comment by Dr15Jones** [comment]
@makortel FYI

**Comment by jpivarski** [comment]
Most of the activity on this issue was in 2017. In 2023, I used the thread-local versus atomic histograms ([referenced above](https://github.com/cms-sw/cmssw/issues/19270#issuecomment-309648812)) as an exercise for students: [ATLAS](https://github.com/jpivarski-talks/2023-11-02-atlas-gpu-python-tutorial/blob/main/projects/project-1-fill-histograms.ipynb), [HSF-India](https://github.com/jpivarski-talks/2023-12-18-hsf-india-tutorial-bhubaneswar/blob/main/lesson-5a-gpu/project-histogram.ipynb). (Although that example is on GPUs, where atomics are a more obvious win than in main memory. In general, the best framework would be one that splits a histogramming task into N locally-filled histograms, to be added together at the end, each of which is filled by M threads with atomics. The ratio of N to M depends on how much memory is available, how likely collisions are between individual bins, etc. Only if you're on an extreme end of the spectrum where M = 1 and un-contested atomics are costly would it be better to ditch the atomics.)

**Comment by makortel** [comment]
> TH2Poly is now supported in DQM as far as I can tell

AFAICT `TH2Poly` support was added in https://github.com/cms-sw/cmssw/pull/47478, but I don't understand what that has to do with the "sparse filling" this issue seems to be about.

I don't see any reference to `THnSparse` in CMSSW (except in `haddnano`). @cms-sw/dqm-l2 Is the `THnSparse` (mentioned in) still in the radar?

**Comment by gabrielmscampos** [comment]
> AFAICT `TH2Poly` support was added in [#47478](https://github.com/cms-sw/cmssw/pull/47478), but I don't understand what that has to do with the "sparse filling" this issue seems to be about.

That is right, it was implemented last year and is supported in the DQM GUI. There are some ongoing experiments with HGCAL using TH2Poly.

> I don't see any reference to `THnSparse` in CMSSW (except in `haddnano`). [@cms-sw/dqm-l2](https://github.com/orgs/cms-sw/teams/dqm-l2) Is the `THnSparse` (mentioned in) still in the radar?

This is definitely not in our radar. I also asked the previous DQM L2 that was taking care of the PR validation, and he hasn't heard about any tasks to add `THnSparse` support. Is this still required? We are in the process of designing a skeleton specification for DQM Phase2, so that might be a good time to think about it if needed.
